sdk/monitor/opentelemetry-exporter/src/export/exporter.ts

```diff
@@ -0,0 +1,150 @@
+    const envelopes = this._applyTelemetryProcessors(payload);
+    this._logger.info(`Exporting ${envelopes.length} envelope(s)`);
+    this._sender.send(envelopes, (err, statusCode, resultString) => {
+      const persistCb = (persistErr: Error | null, persistSuccess?: boolean) => {
+        if (persistErr || !persistSuccess) {
+          return resultCallback(ExportResult.FAILED_NOT_RETRYABLE);
+        }
+        return resultCallback(ExportResult.FAILED_RETRYABLE);
+      };
+
+      if (err) {
+        // Request failed -- always retry
+        this._logger.error(err.message);
+        this._persister.push(envelopes, persistCb);
+      } else if (statusCode === 200) {
+        // Success -- @todo: start retry timer
+        if (!this._retryTimer) {
+          this._retryTimer = setTimeout(() => {
+            this._retryTimer = null;
+            this._sendFirstPersistedFile();
+          }, this._options.batchSendRetryIntervalMs);
+          this._retryTimer.unref();
+        }
+        resultCallback(ExportResult.SUCCESS);
+      } else if (statusCode && isRetriable(statusCode)) {
+        // Failed -- persist failed data
+        if (resultString) {
+          this._logger.info(resultString);
+          const breezeResponse = JSON.parse(resultString) as BreezeResponse;
+          const filteredEnvelopes = breezeResponse.errors.reduce(
+            (acc, v) => [...acc, envelopes[v.index]],
+            [] as Envelope[],
+          );
+          // calls resultCallback(ExportResult) based on result of persister.push
+          this._persister.push(filteredEnvelopes, persistCb);
+        } else {
+          // calls resultCallback(ExportResult) based on result of persister.push
+          this._persister.push(envelopes, persistCb);
+        }
+      } else {
+        // Failed -- not retriable
+        resultCallback(ExportResult.FAILED_NOT_RETRYABLE);
+      }
+    });
+  }
+
+  addTelemetryProcessor(processor: TelemetryProcessor): void {
+    this._telemetryProcessors.push(processor);
+  }
+
+  clearTelemetryProcessors() {
+    this._telemetryProcessors = [];
+  }
+
+  shutdown() {
+    this._sender.shutdown();
+  }
+
+  protected _applyTelemetryProcessors(envelopes: Envelope[]): Envelope[] {
+    const filteredEnvelopes: Envelope[] = [];
+    envelopes.forEach((envelope) => {
+      let accepted = true;
+
+      this._telemetryProcessors.forEach((processor) => {
+        // Don't use CPU cycles if item is already rejected
+        if (accepted && processor(envelope) === false) {
+          accepted = false;
+        }
+      });
+
+      if (accepted) {
+        filteredEnvelopes.push(envelope);
+      }
+    });
+
+    return filteredEnvelopes;
+  }
+
+  private _sendFirstPersistedFile() {
+    this._persister.shift((err, envelopes) => {
+      if (err) {
+        this._logger.warn(`Failed to fetch persisted file`, err);
+      } else if (envelopes) {
+        this._sender.send(envelopes, () => {
+          /** no-op */
```

Comment:
having to pass a noop callback feels a little silly, could we make that optional? Or if we used promises we wouldn't need to worry... ;)

Reply:
Changed to optional. Will file todo to transition to async/await